File: deno/cargo_tracker/sample1/cargo/cargo.ts

import { UnLocode, RouteSpecification, Itinerary, LocationTime } from '../common.ts'

export type { UnLocode, RouteSpecification, Itinerary, LocationTime } from '../common.ts'

export type TrackingId = string

export type CargoEvent = CargoCreated | DestinationChanged | DeadlineChanged | 
    RouteAssigned | Closed

export interface CargoCreated {
    tag: 'cargo-event.created'
    trackingId: TrackingId
    routeSpec: RouteSpecification
}

export interface DestinationChanged {
    tag: 'cargo-event.destination-changed'
    trackingId: TrackingId
    newDestination: UnLocode
}

export interface DeadlineChanged {
    tag: 'cargo-event.deadline-changed'
    trackingId: TrackingId
    newDeadline: Date
}

export interface RouteAssigned {
    tag: 'cargo-event.route-assigned'
    trackingId: TrackingId
    itinerary: Itinerary
}

export interface Closed {
    tag: 'cargo-event.closed'
    trackingId: TrackingId
}

export type Cargo = NoneCargo | ActiveCargo
export type ActiveCargo = UnroutedCargo | RoutedCargo | MisroutedCargo | ClosedCargo

export interface NoneCargo {
    tag: 'cargo.none'
}

export interface UnroutedCargo {
    tag: 'cargo.unrouted'
    trackingId: TrackingId
    routeSpec: RouteSpecification
}

export interface RoutedCargo {
    tag: 'cargo.routed'
    trackingId: TrackingId
    routeSpec: RouteSpecification
    itinerary: Itinerary
}

export interface MisroutedCargo {
    tag: 'cargo.misrouted'
    trackingId: TrackingId
    routeSpec: RouteSpecification
    itinerary: Itinerary
}

export interface ClosedCargo {
    tag: 'cargo.closed'
    trackingId: TrackingId
    routeSpec: RouteSpecification
    itinerary: Itinerary
}

export type CargoAction = (state: ActiveCargo) => CargoResult
type CargoResult = { cargo: ActiveCargo, event: CargoEvent } | undefined
type CargoOptional = ActiveCargo | undefined

const noneCargo: NoneCargo = { tag: 'cargo.none' }

export function create(trackingId: TrackingId, origin: UnLocode, destination: UnLocode, 
    arrivalDeadline: Date): CargoResult {

    if (!isFuture(arrivalDeadline)) {
        return undefined
    }

    const routeSpec = { origin, destination, arrivalDeadline }

    return applyEvent(noneCargo, {
        tag: 'cargo-event.created',
        trackingId,
        routeSpec
    })
}

export function changeDestination(destination: UnLocode): CargoAction {
    return state => {
        if (state.routeSpec.destination == destination) {
            return undefined
        }

        return applyEvent(state, {
            tag: 'cargo-event.destination-changed',
            trackingId: state.trackingId,
            newDestination: destination
        })
    }
}

export function changeDeadline(deadline: Date): CargoAction {
    return state => {
        if (!isFuture(deadline)) {
            return undefined
        }

        if (state.routeSpec.arrivalDeadline.getTime() == deadline.getTime()) {
            return undefined
        }

        return applyEvent(state, {
            tag: 'cargo-event.deadline-changed',
            trackingId: state.trackingId,
            newDeadline: deadline
        })
    }
}

export function assignToRoute(itinerary: Itinerary): CargoAction {
    return state => {
        return applyEvent(state, {
            tag: 'cargo-event.route-assigned',
            trackingId: state.trackingId,
            itinerary
        })
    }
}

export function close(state: ActiveCargo): CargoResult {
    return applyEvent(state, {
        tag: 'cargo-event.closed',
        trackingId: state.trackingId
    })
}

function applyEvent(state: Cargo, event: CargoEvent): CargoResult {
    const cargo = transit(state, event)

    if (cargo) {
        return { cargo, event }
    }

    return undefined
}

function transit(state: Cargo, event: CargoEvent): CargoOptional {
    switch (state.tag) {
        case 'cargo.closed':
            return undefined
        case 'cargo.none':
            if (event.tag != 'cargo-event.created') {
                return undefined
            }

            return {
                tag: 'cargo.unrouted',
                trackingId: event.trackingId,
                routeSpec: event.routeSpec
            }
        case 'cargo.unrouted':
            if (state.trackingId != event.trackingId) {
                return undefined
            }
            return transitForUnrouted(state, event)
        case 'cargo.routed':
            if (state.trackingId != event.trackingId) {
                return undefined
            }
            return transitForRouted(state, event)
        case 'cargo.misrouted':
            if (state.trackingId != event.trackingId) {
                return undefined
            }
            return transitForRouted(state, event)
    }
}

function transitForUnrouted(state: UnroutedCargo, event: CargoEvent): CargoOptional {
    switch (event.tag) {
        case 'cargo-event.created':
        case 'cargo-event.closed':
            return undefined
        case 'cargo-event.destination-changed':
            return {
                ...state,
                routeSpec: { ...state.routeSpec, destination: event.newDestination }
            }            
        case 'cargo-event.deadline-changed':
            return {
                ...state,
                routeSpec: { ...state.routeSpec, arrivalDeadline: event.newDeadline }
            }
        case 'cargo-event.route-assigned':
            return inspectRoute({
                ...state,
                tag: 'cargo.routed',
                itinerary: event.itinerary
            })
    }
}

function transitForRouted(state: RoutedCargo | MisroutedCargo, event: CargoEvent): CargoOptional {
    switch (event.tag) {
        case 'cargo-event.created':
            return undefined
        case 'cargo-event.closed':
            return { ...state, tag: 'cargo.closed' }
        case 'cargo-event.destination-changed':
            return inspectRoute({
                ...state, 
                routeSpec: { ...state.routeSpec, destination: event.newDestination }
            })
        case 'cargo-event.deadline-changed':
            return inspectRoute({ 
                ...state, 
                routeSpec: { ...state.routeSpec, arrivalDeadline: event.newDeadline } 
            })
        case 'cargo-event.route-assigned':
            return inspectRoute({ 
                ...state,
                itinerary: event.itinerary
            })
    }
}

function inspectRoute(state: RoutedCargo | MisroutedCargo): ActiveCargo {
    const ok = isSatisfiedWithRoute(state.itinerary, state.routeSpec)

    if (ok && state.tag == 'cargo.misrouted') {
        return { ...state, tag: 'cargo.routed' }
    }

    if (!ok && state.tag == 'cargo.routed') {
        return { ...state, tag: 'cargo.misrouted'}
    }

    return state
}

const isFuture = (date: Date) => date.getTime() > Date.now()

const isSatisfiedWithRoute = (itinerary: Itinerary, routeSpec: RouteSpecification) => {
    const r = itinerary.legs.reduce((acc, v) => {
        const { location, time } = v.load

        if (acc && location == acc.location && time.getTime() >= acc.time.getTime()) {
            return v.unload
        }

        return undefined
    }, { location: routeSpec.origin, time: new Date(0) } as LocationTime | undefined)

    return r && r.location == routeSpec.destination && 
        r.time.getTime() < routeSpec.arrivalDeadline.getTime()
}

File: scala/cargo-tracker/model2/src/main/scala/cargo.scala
package cargo

import cats.{Id, ~>}
import cats.arrow.FunctionK
import cats.free.Free
import Free.liftF
import cats.data.State

import java.time.LocalDateTime

type TrackingId = String
type UnLocode = String
type VoyageNo = String
type Date = LocalDateTime

case class RouteSpecification(origin: UnLocode, destination: UnLocode, deadline: Date)

case class Itinerary(legs: List[Leg])

case class LocationTime(location: UnLocode, time: Date)

case class Leg(voyageNo: VoyageNo, load: LocationTime, unload: LocationTime)

trait HasRouteSpec {
  val trackingId: TrackingId
  val routeSpec: RouteSpecification
}

trait HasItinerary extends HasRouteSpec {
  val itinerary: Itinerary
}

enum Cargo:
  case Empty()
  case Unrouted(trackingId: TrackingId, routeSpec: RouteSpecification) extends Cargo, HasRouteSpec
  case Routed(trackingId: TrackingId, routeSpec: RouteSpecification, itinerary: Itinerary) extends Cargo, HasItinerary
  case Misrouted(trackingId: TrackingId, routeSpec: RouteSpecification, itinerary: Itinerary) extends Cargo, HasItinerary
  case Closed(trackingId: TrackingId, routeSpec: RouteSpecification, itinerary: Itinerary) extends Cargo, HasItinerary

enum Event:
  case Created(trackingId: TrackingId, routeSpec: RouteSpecification)
  case AssignedRoute(trackingId: TrackingId, itinerary: Itinerary)
  case ChangedDestination(trackingId: TrackingId, newDestination: UnLocode)
  case ChangedDeadline(trackingId: TrackingId, newDeadline: Date)
  case Closed(trackingId: TrackingId)

enum CommandA[A]:
  case Create(trackingId: TrackingId, routeSpec: RouteSpecification) extends CommandA[List[Event]]
  case AssignRoute(itinerary: Itinerary) extends CommandA[List[Event]]
  case ChangeDestination(newDestination: UnLocode) extends CommandA[List[Event]]
  case ChangeDeadline(newDeadline: Date) extends CommandA[List[Event]]
  case Close() extends CommandA[List[Event]]

type Command[A] = Free[CommandA, A]

import CommandA.*

type CargoState[A] = State[Cargo, A]

object CargoAction:
  def emptyCargo = Cargo.Empty()

  def create(trackingId: TrackingId, origin: UnLocode, destination: UnLocode, deadline: Date) =
    liftF(Create(trackingId, RouteSpecification(origin, destination, deadline)))

  def assignRoute(itinerary: Itinerary) = liftF(AssignRoute(itinerary))

  def changeDestination(destination: UnLocode) = liftF(ChangeDestination(destination))

  def changeDeadline(deadline: Date) = liftF(ChangeDeadline(deadline))

  def close() = liftF(Close())

  def interpret: CommandA ~> CargoState =
    new (CommandA ~> CargoState) {
      def apply[A](cmd: CommandA[A]): CargoState[A] =
        cmd match {
          case Create(t, r) => State {
            case Cargo.Empty() if t.nonBlank && r.deadline.isFuture =>
              (Cargo.Unrouted(t, r), List(Event.Created(t, r)))
            case s => (s, List.empty[Event])
          }
          case AssignRoute(i) => State {
            case s: (Cargo.Unrouted | Cargo.Routed | Cargo.Misrouted) if i.legs.nonEmpty =>
              (
                inspectRoute(Cargo.Routed(s.trackingId, s.routeSpec, i)),
                List(Event.AssignedRoute(s.trackingId, i))
              )
            case s => (s, List.empty[Event])
          }
          case ChangeDestination(d) => State {
            case s: Cargo.Unrouted =>
              (
                s.copy(routeSpec = s.routeSpec.copy(destination = d)),
                List(Event.ChangedDestination(s.trackingId, d))
              )
            case s: (Cargo.Routed | Cargo.Misrouted) =>
              (
                inspectRoute(Cargo.Routed(s.trackingId, s.routeSpec.copy(destination = d), s.itinerary)),
                List(Event.ChangedDestination(s.trackingId, d))
              )
            case s => (s, List.empty[Event])
          }
          case ChangeDeadline(d) => State {
            case s: Cargo.Unrouted if d.isFuture =>
              (
                s.copy(routeSpec = s.routeSpec.copy(deadline = d)),
                List(Event.ChangedDeadline(s.trackingId, d))
              )
            case s: (Cargo.Routed | Cargo.Misrouted) if d.isFuture =>
              (
                inspectRoute(Cargo.Routed(s.trackingId, s.routeSpec.copy(deadline = d), s.itinerary)),
                List(Event.ChangedDeadline(s.trackingId, d))
              )
            case s => (s, List.empty[Event])
          }
          case Close() => State {
            case s: (Cargo.Routed | Cargo.Misrouted) =>
              (Cargo.Closed(s.trackingId, s.routeSpec, s.itinerary), List(Event.Closed(s.trackingId)))
            case s => (s, List.empty[Event])
          }
        }
    }
  end interpret

  private def isSatisfiedWithRoute(itinerary: Itinerary, routeSpec: RouteSpecification): Boolean =
    val fst: Option[LocationTime] = Some(LocationTime(routeSpec.origin, LocalDateTime.MIN))

    val lst = itinerary.legs.foldLeft(fst)((acc, leg) =>
      for {
        c <- acc
        if c.location == leg.load.location && c.time.isBeforeOrEq(leg.load.time)
      } yield leg.unload
    )

    lst.exists(lt => lt.location == routeSpec.destination && lt.time.isBefore(routeSpec.deadline))

  private def inspectRoute(state: Cargo.Routed): Cargo =
    if (isSatisfiedWithRoute(state.itinerary, state.routeSpec))
      state
    else
      Cargo.Misrouted(state.trackingId, state.routeSpec, state.itinerary)

  extension (s: String)
    def nonBlank: Boolean = !s.isBlank

  extension (d: Date)
    def isFuture: Boolean = d.isAfter(LocalDateTime.now())
    def isBeforeOrEq(t: Date): Boolean = d.compareTo(t) <= 0

end CargoAction